Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp

```diff
@@ -83,4 +76,43 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
     return StorageNamespaceImpl::createTransientLocalStorageNamespace(m_localStorageNamespaceIdentifier, topLevelOrigin, quota);
 }
 
+RefPtr<WebCore::StorageNamespace> WebStorageNamespaceProvider::sessionStorageNamespace(WebCore::SecurityOrigin& topLevelOrigin, WebCore::Page& page, ShouldCreateIfNotExists createIfNotExists)
+{
+    ASSERT(m_sessionStorageQuota != WebCore::StorageMap::noQuota);
+
+    auto& webPage = WebPage::fromCorePage(page);
+
+    // The identifier of a session storage namespace is the WebPageProxyIdentifier. It is possible we have several WebPage objects in a single process for the same
+    // WebPageProxyIdentifier and these need to share the same namespace instance so we know where to route the IPC to.
+    auto& namespacesSlot = m_sessionStorageNamespaces.add(webPage.sessionStorageNamespaceIdentifier(), SessionStorageNamespaces { }).iterator->value;
+    auto& slot = namespacesSlot.map.add(topLevelOrigin.data(), nullptr).iterator->value;
+    if (!slot && createIfNotExists == ShouldCreateIfNotExists::Yes)
+        slot = StorageNamespaceImpl::createSessionStorageNamespace(webPage.sessionStorageNamespaceIdentifier(), webPage.identifier(), topLevelOrigin, m_sessionStorageQuota).ptr();
+    return slot;
+}
+
+void WebStorageNamespaceProvider::setSessionStorageNamespace(WebCore::SecurityOrigin& topLevelOrigin, WebCore::Page& page, RefPtr<WebCore::StorageNamespace>&& newNamespace)
+{
+    auto& webPage = WebPage::fromCorePage(page);
+
+    auto& namespacesSlot = m_sessionStorageNamespaces.add(webPage.sessionStorageNamespaceIdentifier(), SessionStorageNamespaces { }).iterator->value;
+    auto& slot = namespacesSlot.map.add(topLevelOrigin.data(), nullptr).iterator->value;
+    slot = WTFMove(newNamespace);
+}
+
+void WebStorageNamespaceProvider::incrementUseCount(StorageNamespaceImpl::Identifier identifier)
+{
+    auto& slot = m_sessionStorageNamespaces.add(identifier, SessionStorageNamespaces { }).iterator->value;
```

Comment:
Do we expect the identifier to be already in the map? If so, it would be good to use find() instead of add() and ASSERT() that the identifier is indeed in the map.